activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -193,6 +205,80 @@ def wait_poll(timeout)
         end
       end
 
+      # Adds the ability to turn a basic fair FIFO queue into one
+      # biased to some thread.
+      module BiasableQueue # :nodoc:
+        class BiasedConditionVariable
```

Comment:
Sorry, I've commented earlier but it looks like it never reached Github. I've pulled the rails source and tried to understand this class. I have to admit I was not able to grasp it. I know what Java uses "Biased locking" to optimize. I assume this is related but I am not sure about the conditions when this class can be used etc. @thedarkone could you provide more documentation for this class please.